gprofiler/metadata/system_metadata.py

```diff
@@ -0,0 +1,236 @@
+import array
+import fcntl
+import os
+import platform
+import re
+import socket
+import struct
+import subprocess
+import sys
+import time
+from dataclasses import dataclass
+from typing import Dict, Optional, Tuple
+
+import distro  # type: ignore
+import psutil
+
+from gprofiler.log import get_logger_adapter
+from gprofiler.utils import run_in_ns, run_process
+
+logger = get_logger_adapter(__name__)
+hostname: Optional[str] = None
+RUN_MODE_TO_DEPLOYMENT_TYPE: Dict[str, str] = {
+    "k8s": "k8s",
+    "container": "containers",
+    "standalone_executable": "instances",
+    "local_python": "instances",
+}
+
+
+def get_libc_version() -> Tuple[str, str]:
+    # platform.libc_ver fails for musl, sadly (produces empty results).
+    # so we'll run "ldd --version" and extract the version string from it.
+    # not passing "encoding"/"text" - this runs in a different mount namespace, and Python fails to
+    # load the files it needs for those encodings (getting LookupError: unknown encoding: ascii)
+    def decode_libc_version(version: bytes) -> str:
+        return version.decode("utf-8", errors="replace")
+
+    ldd_version = run_process(
+        ["ldd", "--version"], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, suppress_log=True, check=False
+    ).stdout
+    # catches GLIBC & EGLIBC
+    m = re.search(br"GLIBC (.*?)\)", ldd_version)
+    if m is not None:
+        return "glibc", decode_libc_version(m.group(1))
+    # catches GNU libc
+    m = re.search(br"\(GNU libc\) (.*?)\n", ldd_version)
+    if m is not None:
+        return "glibc", decode_libc_version(m.group(1))
+    # musl
+    m = re.search(br"musl libc.*?\nVersion (.*?)\n", ldd_version, re.M)
+    if m is not None:
+        return "musl", decode_libc_version(m.group(1))
+
+    return "unknown", decode_libc_version(ldd_version)
+
+
+def get_run_mode() -> str:
+    if os.getenv("GPROFILER_IN_K8S") is not None:  # set in k8s/gprofiler.yaml
+        return "k8s"
+    elif os.getenv("GPROFILER_IN_CONTAINER") is not None:  # set by our Dockerfile
+        return "container"
+    elif os.getenv("STATICX_BUNDLE_DIR") is not None:  # set by staticx
+        return "standalone_executable"
+    else:
+        return "local_python"
+
+
+def get_deployment_type(run_mode: str) -> str:
+    return RUN_MODE_TO_DEPLOYMENT_TYPE.get(run_mode, "unknown")
+
+
+def get_private_ip() -> str:
+    # Fetches the local IP. Attempts to get it locally. If it fails, it will attempt to fetch it by connecting to
```

Comment:
Either case it's locally. The "connect" to 8.8.8.8 is UDP - this just tells Linux to figure out which local endpoint is going to be used to communicate with `8.8.8.8:53`; no packet is sent.